### ISSUE TITLE
Make ImplicitUsings does not work out of the box with .NET Framework project

### ISSUE THREAD

**Issue by madelson** [dotnet/sdk#24909]
*Apologies if this is not the right repo! Reporting here because there are at least some issues in this repo referencing `<ImplicitUsings>`*

### Is your feature request related to a problem? Please describe.

Whenever I create a new class library project, it defaults to setting `<ImplicitUsings>enable</ImplicitUsings>`. This is good. However, when I multi-target or re-target the project to hit .NET Framework, I get the following compiler error:

> The type or namespace name 'Http' does not exist in the namespace 'System.Net' (are you missing an assembly reference?)

This can be resolved by adding a framework reference, but it would be nice if it "just worked".

### Describe the solution you'd like

I'd like for this to "just work" out of the box. One easy way to do this would be to change the generated usings file (ProjectName.GlobalUsings.g.cs) to look like the following:

```cs
// <auto-generated/>
global using global::System;
global using global::System.Collections.Generic;
global using global::System.IO;
global using global::System.Linq;
global using global::System.Net.Http;
global using global::System.Threading;
global using global::System.Threading.Tasks;

// new addition
#if NETFRAMEWORK
namespace System.Net.Http { }
#endif
```

Potentially we could define all the referenced namespaces to ensure that the auto-generated code never gives an error regardless of what framework is targeted and what references are imported.

### Additional context

Example .csproj to reproduce the issue:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net472;netstandard2.0;net6.0</TargetFrameworks>
    <ImplicitUsings>enable</ImplicitUsings>
    <LangVersion>10</LangVersion>
  </PropertyGroup>

</Project>
```

**Comment by baronfel** [comment]
At minimum, we should condition the Using for System.Net.Http so that it is only added for TFMs that natively include the reference to System.Net.Http. Perhaps it's safe enough to use .NET Standard2.0+, .Net Core 3.1, and .Net 5+?

**Comment by madelson** [comment]
@baronfel won't that be a breaking change for existing projects that are using `ImplicitUsings` and have added the System.Net.Http reference?

What is the downside of declaring empty namespaces in the generated code? Feels like that would be a robust way of ensuring that this feature never causes a compilation error.

**Comment by JoshMcCullough** [comment]
> What is the downside of declaring empty namespaces in the generated code?

This would be an ugly hack, for sure.

Is there an appropriate workaround for this? I tried adding my own `GlobalUsings.cs` to the root (alongside the `csproj`), but the one for .NET 4.8 is still generated.

**Comment by madelson** [comment]
> That would be an ugly hack, for sure.

@JoshMcCullough curious why you see this as an ugly hack if it just affects the auto-generated code.

> Is there an appropriate workaround for this?

You can either do this in the csproj either by removing the `<Using Remove="..." />` (see https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#using) or you can add the corresponding framework reference to bring in an assembly that includes the namespace.

**Comment by JoshMcCullough** [comment]
> curious why you see this as an ugly hack if it just affects the auto-generated code.

@madelson because it's a hack! It's a workaround rather than solving the actual problem which is, that those `global using` statements shouldn't have been added in the first place since the library isn't included as a dependency.

I happen to think even generated code should be beautiful and concise.

**Comment by madelson** [comment]
> solving the actual problem which is, that those global using statements shouldn't have been added in the first place since the library isn't included as a dependency

The challenge here is that in .NET namespaces have a many to many relationship with libraries so the only way to know for sure whether a namespace will be available at compile time is to examine all the types in all the imported libraries. We could avoid this by making assumptions about what libraries are likely to be required for each namespace to work... but that also feels like a hack.

**Comment by JoshMcCullough** [comment]
> ...but that also feels like a hack.

Agreed. So like most issues, the solution is not plainly clear!

**Comment by tqk2811** [comment]
lol, i take alot hours with project not using http request but error at `system.net.http`

**Comment by madelson** [comment]
Any chance this will be addressed (either via my PR or another means) soon? I just ran into this again with another project; so annoying to have to add
```
  <ItemGroup>
    <Using Remove="System.Net.Http"/>
  </ItemGroup>
```
to every csproj file when the point of implicit usings is to get rid of boilerplate.

**Comment by dsplaisted** [comment]
Rather than adding empty namespaces, I think what we probably want to do here is add conditions to where they are added to MSBuild:

```xml
<Using Include="System.Net.Http" Condition="'$(TargetFrameworkIdentifier)' != '.NETFramework'"/>
```

**Comment by martincostello** [comment]
> @baronfel won't that be a breaking change for existing projects that are using `ImplicitUsings` and have added the System.Net.Http reference?

I just found this issue after a library I maintain where we do multi-targeting had the build break when I updated to the .NET 8 preview 5 SDK, so I would say it is breaking yes 😄

**Comment by baronfel** [comment]
Yes, thanks for bringing this to my attention. We should attempt to dedupe these items as a mitigation.

**Comment by JL03-Yue** [comment]
resolved in #32630
